src/main/java/org/broadinstitute/dsde/consent/ontology/datause/DataUseMatchCasesV3.java

```diff
@@ -0,0 +1,276 @@
+package org.broadinstitute.dsde.consent.ontology.datause;
+
+import org.broadinstitute.dsde.consent.ontology.model.DataUseV3;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public class DataUseMatchCasesV3 {
+  private static final String HMB_F1 = "The GRU Research Purpose does not match the HMB data use limitations.";
+  private static final String HMB_F2 = "The HMB Research Purpose does not match the Disease-Specific data use limitations.";
+  private static final String DS_F1 = "The unspecified Disease Research Purpose does not match the Disease-Specific data use limitations.";
+  private static final String DS_F2 = "The Disease-Specific: %s Research Purpose is not a valid subclass of the Disease-Specific data use limitations.";
+  private static final String MDS_F1 = "The Methods development Research Purpose does not match the Disease-Specific data use limitations.";
+  private static final String POA_F1 = "The Populations, Origins, Ancestry Research Purpose does not match the HMB or Disease-Specific data use limitation.";
+  private static final String NCU_F1 = "The Commercial Use Research Purpose does not match the No Commercial Use data use limitation.";
+  private static final String Abstain = "The Research Purpose does not result in DUOS Decision.";
+
+  /**
+   * RP: Disease Focused Research
+   * Future use is limited to research involving the following disease area(s) [DS]
+   * Datasets:
+   *      Any dataset with GRU=true
+   *      Any dataset with HMB=true
+   *      Any dataset tagged to this disease exactly
+   *      Any dataset tagged to a DOID ontology Parent of disease X
+   *
+   * @param purpose The data use object representing the Research Purpose
+   * @param dataset The data use object representing the Dataset
+   * @param purposeDiseaseIdMap is a map of each purpose term id to a list of that term's parent term ids
+   */
+  static MatchResult matchDiseases(
+      DataUseV3 purpose, DataUseV3 dataset, Map<String, List<String>> purposeDiseaseIdMap) {
+    // short-circuit if no disease focused research
+    if (purpose.getDiseaseRestrictions().isEmpty() && dataset.getDiseaseRestrictions().isEmpty()) {
+      return MatchResult.from(MatchResultType.APPROVE, Collections.emptyList());
+    }
+    // short-circuit if dataset is GRU
```

Comment:
Minor suggestion for comments here and throughout. In general, we're short-circuiting logic checks in the cases where there are no logical conditions to check. In cases where there ARE logical conditions to check, we're no longer short-circuiting. Instead, we're following the logic as outlined in the logic table.
```suggestion
    // Approve all GRU conditions
```